### ISSUE TITLE
HTTP-Adapter: IllegalStateException: Response has already been written

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#2892]
The HTTP adapter showed this error in the logs:
```
DEBUG VertxBasedHttpProtocolAdapter  failed to send http response for [EVENT] message from device [tenantId: test, deviceId: test_123]: response already closed
WARN  JaegerSpan  Span has already been finished; will not be reported again.
WARN  TracingHelper  An unexpected error occurred!
java.lang.IllegalStateException: Response has already been written
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:394)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:382)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:451)
	at org.eclipse.hono.adapter.http.AbstractVertxBasedHttpProtocolAdapter.lambda$doUploadMessage$31(AbstractVertxBasedHttpProtocolAdapter.java:738)
```

**Comment by calohmn** [comment]
I thought it was easy to reproduce (cancelling client request before HTTP adapter sends response), but it wasn't.

**Comment by suzanaAngelovaIvanova** [comment]
I would like to work on this issue.

**Comment by sophokles73** [comment]
Sure :-)

**Comment by suzanaAngelovaIvanova** [comment]
Hi @sophokles73, I'm trying to reproduce the bug, but I can't, are there any steps that I can follow?

**Comment by sophokles73** [comment]
I guess you will need to try to correlate the stack trace with the code that writes the response to a device's request in the [HTTP adapter](https://github.com/eclipse/hono/blob/857511a6899249fedf536bf0a81c971301bda5b8/adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java#L579).

**Comment by suzanaAngelovaIvanova** [comment]
CASE 1: curl -i -u $MY_DEVICE@$MY_TENANT:$MY_PWD -H 'Content-Type: application/json' -H "hono-ttd: 60" --data-binary '{"alarm": "fire"}' http://$HTTP_ADAPTER_IP:8080/event -v

CASE 2: curl -i -u $MY_DEVICE@$MY_TENANT:$MY_PWD -H 'Content-Type: application/json' --data-binary '{"alarm": "fire"}' http://$HTTP_ADAPTER_IP:8080/event -v

Hi, I tried to reproduce it by adding to the request the following property **hono-ttd** (CASE 1)
and I canceled the request before the **hono-ttd** expires, the result is that it does not enter into Line: 699 (.map(proceed -> {.....)


I am adding a log.
**curl -i -u $MY_DEVICE@$MY_TENANT:$MY_PWD -H 'Content-Type: application/json' -H "hono-ttd: 60" --data-binary '{"alarm": "fire"}' http://$HTTP_ADAPTER_IP:8080/event -v**

2022-07-06 08:08:42,211 TRACE [org.ecl.hon.cli.kaf.con.AsyncHandlingAutoCommitKafkaConsumer] (vert.x-kafka-consumer-thread-1) skip periodic commit - no offsets to commit
2022-07-06 08:08:45,472 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) protocol adapter [hono-http] is enabled for tenant [my-tenant2]
2022-07-06 08:08:45,472 DEBUG [org.ecl.hon.ada.aut.dev.UsernamePasswordAuthProvider] (executor-thread-0) validating password hash on vert.x worker thread [executor-thread-0]
2022-07-06 08:08:45,594 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) protocol adapter [hono-http] is enabled for tenant [my-tenant2]
2022-07-06 08:08:45,600 TRACE [org.ecl.hon.ada.res.DataVolumeAsyncCacheLoader] (vert.x-eventloop-thread-0) no resource limits configured for tenant [my-tenant2]
**2022-07-06 08:08:45,610 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs before ttdTracker**
**2022-07-06 08:08:45,610 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) **SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs before commandConsumerTracker** 
**2022-07-06 08:08:45,610 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS Line 1054**
2022-07-06 08:08:45,610 TRACE [org.ecl.hon.cli.com.CommandRouterCommandConsumerFactory] (vert.x-eventloop-thread-0) create command consumer [tenant-id: my-tenant2, device-id: my-device2, gateway-id: null]
2022-07-06 08:08:45,610 DEBUG [org.ecl.hon.cli.uti.CachingClientFactory] (vert.x-eventloop-thread-0) reusing cached client [cmd_router-my-tenant2]
2022-07-06 08:08:45,611 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) sent request [target address: cmd_router/my-tenant2, subject: register-command-consumer, correlation ID: cmd_router-client-a71956a4-4c62-48af-a2ec-d6960ae465bf, device ID: my-device2] to service
**2022-07-06 08:08:45,611 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS 1073**
2022-07-06 08:08:45,631 TRACE [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) service has accepted request [target address: cmd_router/my-tenant2, subject: register-command-consumer, correlation ID: cmd_router-client-a71956a4-4c62-48af-a2ec-d6960ae465bf]
2022-07-06 08:08:45,633 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) received response [reply-to: cmd_router/my-tenant2/49312ffa-9393-425f-885f-8581516b2be7, subject: null, correlation ID: cmd_router-client-a71956a4-4c62-48af-a2ec-d6960ae465bf, status: 204, cache-directive: null]
2022-07-06 08:08:45,635 TRACE [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) adding command reception timer [id: 1180]
**2022-07-06 08:08:45,635 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs IN .compose(ok -> ......** 
2022-07-06 08:08:45,636 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0)  if (messageConsumer != null && !ctx.response().closed() && !ctx.response().ended()) {...... 
2022-07-06 08:08:45,636 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) sending event [tenantId: my-tenant2, deviceId: my-device2, contentType: application/json, properties: {orig_adapter=hono-http, qos=1, ttd=60, orig_address=/event}]
2022-07-06 08:08:45,637 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) sending message to Kafka [topic: hono.event.my-tenant2, tenantId: my-tenant2, deviceId: my-device2]
2022-07-06 08:08:45,638 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) producing message [topic: hono.event.my-tenant2, key: my-device2, partition: null, timestamp: null, headers: {content-type=application/json,orig_adapter=hono-http,orig_address=/event,qos=1,device_id=my-device2,ttd=60,creation-time=1657094925637,traceparent=00-925db6b8400781b381d9ed8cecb54d65-47d2a5506f68b6bf-01}]
2022-07-06 08:08:45,640 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) client provided response handler did not settle message, auto-accepting ...
2022-07-06 08:08:45,641 TRACE [org.ecl.hon.cli.amq.con.imp.HonoConnectionImpl] (vert.x-eventloop-thread-0) handling message [remotely settled: false, queued messages: 0, remaining credit: 199]
2022-07-06 08:08:45,653 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) message produced to Kafka [topic: hono.event.my-tenant2, key: my-device2, partition: 0, offset: 2, timestamp: 1657094925646]
2022-07-06 08:08:47,238 TRACE [org.ecl.hon.cli.kaf.con.AsyncHandlingAutoCommitKafkaConsumer] (vert.x-kafka-consumer-thread-1) skip periodic commit - no offsets to commit
2022-07-06 08:08:51,371 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) device [tenant: my-tenant2, device-id: my-device2] closed connection before response could be sent
2022-07-06 08:08:51,371 TRACE [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) Cancelled timer id 1180
2022-07-06 08:08:51,372 TRACE [org.ecl.hon.cli.com.CommandRouterCommandConsumerFactory] (vert.x-eventloop-thread-0) remove command consumer [tenant-id: my-tenant2, device-id: my-device2]
2022-07-06 08:08:51,372 TRACE [org.ecl.hon.cli.com.CommandHandlers] (vert.x-eventloop-thread-0) Removed CommandHandlerWrapper{tenantId='my-tenant2', deviceId='my-device2'}: true
2022-07-06 08:08:51,373 DEBUG [org.ecl.hon.cli.uti.CachingClientFactory] (vert.x-eventloop-thread-0) reusing cached client [cmd_router-my-tenant2]
2022-07-06 08:08:51,376 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) sent request [target address: cmd_router/my-tenant2, subject: unregister-command-consumer, correlation ID: cmd_router-client-f865f609-14ba-4273-9ab5-6e89fa642489, device ID: my-device2] to service
2022-07-06 08:08:51,390 TRACE [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) service has accepted request [target address: cmd_router/my-tenant2, subject: unregister-command-consumer, correlation ID: cmd_router-client-f865f609-14ba-4273-9ab5-6e89fa642489]
2022-07-06 08:08:51,390 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) received response [reply-to: cmd_router/my-tenant2/49312ffa-9393-425f-885f-8581516b2be7, subject: null, correlation ID: cmd_router-client-f865f609-14ba-4273-9ab5-6e89fa642489, status: 204, cache-directive: null]
2022-07-06 08:08:51,391 TRACE [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) connection got closed before response could be sent
2022-07-06 08:08:51,398 WARN  [io.qua.ope.run.QuarkusContextStorage] (vert.x-eventloop-thread-0) Context in storage not the expected context, Scope.close was not called correctly
2022-07-06 08:08:51,398 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-0) client provided response handler did not settle message, auto-accepting ...
2022-07-06 08:08:51,398 TRACE [org.ecl.hon.cli.amq.con.imp.HonoConnectionImpl] (vert.x-eventloop-thread-0) handling message [remotely settled: false, queued messages: 0, remaining credit: 199]




The second log is when the request is executed without this property(CASE 2) and the response is 202

**curl -i -u $MY_DEVICE@$MY_TENANT:$MY_PWD -H 'Content-Type: application/json' --data-binary '{"alarm": "fire"}' http://$HTTP_ADAPTER_IP:8080/event -v**


2022-07-06 08:07:55,613 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) protocol adapter [hono-http] is enabled for tenant [my-tenant2]
2022-07-06 08:07:55,614 DEBUG [org.ecl.hon.ada.aut.dev.UsernamePasswordAuthProvider] (executor-thread-0) validating password hash on vert.x worker thread [executor-thread-0]
2022-07-06 08:07:55,712 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) protocol adapter [hono-http] is enabled for tenant [my-tenant2]
2022-07-06 08:07:55,714 TRACE [org.ecl.hon.ada.res.DataVolumeAsyncCacheLoader] (vert.x-eventloop-thread-0) no resource limits configured for tenant [my-tenant2]
**2022-07-06 08:07:55,715 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs before ttdTracker** 
**2022-07-06 08:07:55,715 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs before commandConsumerTracker** 
**2022-07-06 08:07:55,716 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs IN .compose(ok -> ......** 
2022-07-06 08:07:55,716 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) sending event [tenantId: my-tenant2, deviceId: my-device2, contentType: application/json, properties: {orig_adapter=hono-http, qos=1, orig_address=/event}]
2022-07-06 08:07:55,718 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) sending message to Kafka [topic: hono.event.my-tenant2, tenantId: my-tenant2, deviceId: my-device2]
2022-07-06 08:07:55,719 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) producing message [topic: hono.event.my-tenant2, key: my-device2, partition: null, timestamp: null, headers: {content-type=application/json,orig_adapter=hono-http,orig_address=/event,qos=1,device_id=my-device2,creation-time=1657094875718,traceparent=00-cd8e533a06ed11f9a9096b82b6dc514a-28d11e53bbdcd593-01}]
2022-07-06 08:07:55,741 TRACE [org.ecl.hon.cli.tel.kaf.KafkaBasedEventSender] (vert.x-eventloop-thread-0) message produced to Kafka [topic: hono.event.my-tenant2, key: my-device2, partition: 0, offset: 1, timestamp: 1657094875720]
**2022-07-06 08:07:55,742 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs IN .compose(proceed -> ......
2022-07-06 08:07:55,742 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSs IN .map(proceed -> ......** 
2022-07-06 08:07:55,743 TRACE [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-0) successfully processed [EVENT] message for device [tenantId: my-tenant2, deviceId: my-device2]
2022-07-06 08:07:57,181 TRACE [org.ecl.hon.cli.kaf.con.AsyncHandlingAutoCommitKafkaConsumer] (vert.x-kafka-consumer-thread-1) skip periodic commit - no offsets to commit

**Comment by calohmn** [comment]
> Hi, I tried to reproduce it by adding to the request the following property **hono-ttd** (CASE 1) and I canceled the request before the **hono-ttd** expires, the result is that it does not enter into Line: 699 (.map(proceed -> {.....)

So that means that the `Future` result from the preceding `compose` block isn't getting succeeded.
In fact, since the `recover` block below apparently isn't entered as well (no `cannot process` message in your logs above), this `Future` isn't completed at all, if the `hono-ttd` request is cancelled by the device.

I would have guessed that this could mean a memory leak, but according to some tests I've done here this doesn't seem to be the case. Nonetheless I think this is something that should be corrected, letting the Future be completed when the request gets closed by the device prematurely.

In any case, the above tests seem to show that reproducing the error described in this issue isn't that easy.
My guess here would be that you would need to send a `hono-ttd` request and actually send a command message in response to that and then, while this block
https://github.com/eclipse/hono/blob/5a3ae23bb73be5a5d123eba7f43447a0b6c7c58e/adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java#L685-L694
is being processed (i.e. while an `unregister-cmd-consumer` request is sent to the Command Router), the client has to cancel the request.
In order to achieve this, it could help to add some delay (in the form of a vert.x timer execution) in the above block when testing this.

**Comment by avgustinmm** [comment]
@calohmn 
when connection is closed as @suzanaAngelovaIvanova  does the close handler set in `setTtdRequestConnectionCloseHandler` cancels the timer but doesn't complete `responseReady` that is completed by timer otherwise 
Then the `CompositeFuture.all` is not completed and thus neither compose nor recover methods is called.
This shall be fixed, probably by completing `responseReady` with failure. 
Maybe its possible to be the case that this problem doesn't lead to memory leak, however.

**Comment by suzanaAngelovaIvanova** [comment]
Тhe error reproduces after I add vert.x timer 60 seconds(in block  .compose(proceed -> .....) and call request with ttd 30 seconds

`curl -i -u $MY_DEVICE@$MY_TENANT:$MY_PWD -H 'Content-Type: application/json' -H "hono-ttd:30" --data-binary '{"alarm": "fire"}' [http://$HTTP_ADAPTER_IP:8080/event](http://%24http_adapter_ip:8080/event) -v`

log:
```
2022-07-11 11:07:49,951 WARN  [io.qua.ope.run.QuarkusContextStorage] (vert.x-eventloop-thread-0) Context in storage not the expected context, Scope.close was not called correctly
2022-07-11 11:07:49,954 WARN  [io.qua.ope.run.QuarkusContextStorage] (vert.x-eventloop-thread-0) Context in storage not the expected context, Scope.close was not called correctly
Listening for transport dt_socket at address: 8000
2022-07-11 11:10:26,105 DEBUG [org.ecl.hon.cli.uti.CachingClientFactory] (vert.x-eventloop-thread-1) reusing cached client [tenant]
2022-07-11 11:10:26,107 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) sent request [target address: tenant, subject: get, correlation ID: tenant-client-917e7112-5371-4c49-a3aa-76af729a52ae] to service
2022-07-11 11:10:26,120 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) received response [reply-to: tenant/722b15e7-8c1a-4c68-ab2e-f52fa155ace0, subject: null, correlation ID: tenant-client-917e7112-5371-4c49-a3aa-76af729a52ae, status: 200, cache-directive: max-age = 60]
2022-07-11 11:10:26,121 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-1) protocol adapter [hono-http] is enabled for tenant [my-tenant2]
2022-07-11 11:10:26,122 DEBUG [org.ecl.hon.cli.uti.CachingClientFactory] (vert.x-eventloop-thread-1) reusing cached client [credentials-my-tenant2]
2022-07-11 11:10:26,124 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) sent request [target address: credentials/my-tenant2, subject: get, correlation ID: credentials-client-83e39f17-b410-47f1-a8e2-49d2d7e79d7a] to service
2022-07-11 11:10:26,124 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) client provided response handler did not settle message, auto-accepting ...
2022-07-11 11:10:26,139 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) received response [reply-to: credentials/my-tenant2/e29642bb-082a-4398-84a8-724190db60e7, subject: null, correlation ID: credentials-client-83e39f17-b410-47f1-a8e2-49d2d7e79d7a, status: 200, cache-directive: max-age = 60]
2022-07-11 11:10:26,140 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) client provided response handler did not settle message, auto-accepting ...
2022-07-11 11:10:26,140 DEBUG [org.ecl.hon.ada.aut.dev.UsernamePasswordAuthProvider] (executor-thread-37) validating password hash on vert.x worker thread [executor-thread-37]
2022-07-11 11:10:26,240 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-1) protocol adapter [hono-http] is enabled for tenant [my-tenant2]
2022-07-11 11:10:26,241 DEBUG [org.ecl.hon.cli.uti.CachingClientFactory] (vert.x-eventloop-thread-1) reusing cached client [cmd_router-my-tenant2]
2022-07-11 11:10:26,242 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) sent request [target address: cmd_router/my-tenant2, subject: register-command-consumer, correlation ID: cmd_router-client-7452863a-115d-4548-af4f-d8a1e2ccd764, device ID: my-device2] to service
2022-07-11 11:10:26,255 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) received response [reply-to: cmd_router/my-tenant2/83763695-a2d5-470b-990c-f497cec9f65a, subject: null, correlation ID: cmd_router-client-7452863a-115d-4548-af4f-d8a1e2ccd764, status: 204, cache-directive: null]
2022-07-11 11:10:26,256 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) client provided response handler did not settle message, auto-accepting ...
2022-07-11 11:11:26,110 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-1) device [tenant: my-tenant2, device-id: my-device2] closed connection before response could be sent
2022-07-11 11:11:26,110 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-1) Could not cancel timer id 9834
2022-07-11 11:11:26,110 DEBUG [org.ecl.hon.cli.uti.CachingClientFactory] (vert.x-eventloop-thread-1) reusing cached client [cmd_router-my-tenant2]
2022-07-11 11:11:26,111 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) sent request [target address: cmd_router/my-tenant2, subject: unregister-command-consumer, correlation ID: cmd_router-client-592c41fa-5a6e-47db-bb78-47e5a9d0fd27, device ID: my-device2] to service
2022-07-11 11:11:26,115 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) received response [reply-to: cmd_router/my-tenant2/83763695-a2d5-470b-990c-f497cec9f65a, subject: null, correlation ID: cmd_router-client-592c41fa-5a6e-47db-bb78-47e5a9d0fd27, status: 412, cache-directive: null]
2022-07-11 11:11:26,116 WARN  [io.qua.ope.run.QuarkusContextStorage] (vert.x-eventloop-thread-1) Context in storage not the expected context, Scope.close was not called correctly
2022-07-11 11:11:26,116 DEBUG [org.ecl.hon.cli.amq.RequestResponseClient] (vert.x-eventloop-thread-1) client provided response handler did not settle message, auto-accepting ...
2022-07-11 11:11:56,258 DEBUG [org.ecl.hon.cli.com.CommandRouterCommandConsumerFactory] (vert.x-eventloop-thread-1) command consumer not removed - handler already replaced or removed [tenant: my-tenant2, device: my-device2]
2022-07-11 11:11:56,259 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-1) failed to send http response for [EVENT] message from device [tenantId: my-tenant2, deviceId: my-device2]: response already closed
2022-07-11 11:11:56,260 DEBUG [org.ecl.hon.ada.htt.imp.VertxBasedHttpProtocolAdapter] (vert.x-eventloop-thread-1) cannot process [EVENT] message from device [tenantId: my-tenant2, deviceId: my-device2]: java.lang.IllegalStateException: Response has already been written
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:400)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:388)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:457)
	at org.eclipse.hono.adapter.http.AbstractVertxBasedHttpProtocolAdapter.lambda$doUploadMessage$34(AbstractVertxBasedHttpProtocolAdapter.java:713)
	at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:35)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.core.Promise.tryComplete(Promise.java:121)
	at io.vertx.core.Promise.complete(Promise.java:77)
	at org.eclipse.hono.adapter.http.AbstractVertxBasedHttpProtocolAdapter.lambda$doUploadMessage$27(AbstractVertxBasedHttpProtocolAdapter.java:694)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:893)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:860)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:274)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
	at io.vertx.core.impl.AbstractContext.emit(AbstractContext.java:53)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.run(VertxImpl.java:883)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)

```

**Comment by calohmn** [comment]
Fixed via #3347.
